### ISSUE TITLE
docs: Document how to use aws-sdk-java encryption client with Minio

### ISSUE THREAD

**Issue by harshavardhana** [minio/minio#3736]
## Expected Behavior
Save the additional headers and let the AWS SDK client decipher the encrypted data which was previously uploaded.

## Current Behavior
Currently aws-sdk-java encryption client will not work while decoding the encrypted data.

## Possible Solution
Support saving these headers 

```
     * Encrypted symmetric key header that is used in the Authenticated
     * Encryption (AE) cryptographic module. Older versions of S3 encryption
     * client with encryption-only capability would not be able to recognize
     * this AE key, and therefore will be prevented from mistakenly decrypting
     * ciphertext in AE format.
     */
    public static final String CRYPTO_KEY_V2 = "x-amz-key-v2";

    /** Initialization vector (IV) header that is used in the symmetric and envelope encryption mechanisms */
    public static final String CRYPTO_IV = "x-amz-iv";

    /** JSON-encoded description of encryption materials used during encryption */
    public static final String MATERIALS_DESCRIPTION = "x-amz-matdesc";

    /** Instruction file header to be placed in the metadata of instruction files */
    public static final String CRYPTO_INSTRUCTION_FILE = "x-amz-crypto-instr-file";

    /** Header for the original, unencrypted size of an encrypted object */
    public static final String UNENCRYPTED_CONTENT_LENGTH = "x-amz-unencrypted-content-length";

    /** Header for the optional original unencrypted Content MD5 of an encrypted object */
    public static final String UNENCRYPTED_CONTENT_MD5 = "x-amz-unencrypted-content-md5";
```

## Steps to Reproduce (for bugs)
Follow the aws-sdk-java encryption client document and decipher the data.

## Context
This is necessary since we are expecting users to encrypt and upload.

## Your Environment
master branch, latest aws-sdk-java 1.11.x

**Comment by rakesh-eltropy** [comment]
Thanks Harsha. 

Just to give more details on the use case:

We upload lot of pdf, power points, excel, docs, images and videos using minio java sdk . Along with storing original documents, we convert them into html file so that user can directly view the documents in the browser without downloading. Lot of css and supporting images gets generated and stored. 

We work with banks and financial companies, so security is most important criteria for our product. We would like to store all the documents (original, html and supported css, images) in encrypted format so that system administrators cannot open the files directly.

At the time of rendering html, our application should able to decrypt and display the documents.

I would be happy to provide more details. Its quite urgent for us. Happy to help in closing this ASAP.

**Comment by harshavardhana** [comment]
Since aws-sdk-java uploads with `x-amz-meta` key as prefix it is saved properly on server side and the following program works fine with Minio server.  i.e a putObject uploads using an Asymmetric master key. 

```java
import java.io.ByteArrayInputStream;
import java.io.File;
import java.security.KeyFactory;
import java.security.KeyPair;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.spec.PKCS8EncodedKeySpec;
import java.security.spec.X509EncodedKeySpec;
import java.util.Arrays;
import java.util.Iterator;
import java.util.UUID;

import org.apache.commons.io.FileUtils;
import org.apache.commons.io.IOUtils;
import org.joda.time.DateTime;
import org.joda.time.format.DateTimeFormat;
import org.junit.Assert;

import com.amazonaws.auth.AWSCredentials;
import com.amazonaws.auth.BasicAWSCredentials;
import com.amazonaws.regions.Region;
import com.amazonaws.regions.Regions;
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.AmazonS3EncryptionClient;
import com.amazonaws.services.s3.S3ClientOptions;
import com.amazonaws.services.s3.model.EncryptionMaterials;
import com.amazonaws.services.s3.model.ListVersionsRequest;
import com.amazonaws.services.s3.model.ObjectListing;
import com.amazonaws.services.s3.model.ObjectMetadata;
import com.amazonaws.services.s3.model.PutObjectRequest;
import com.amazonaws.services.s3.model.S3Object;
import com.amazonaws.services.s3.model.S3ObjectSummary;
import com.amazonaws.services.s3.model.S3VersionSummary;
import com.amazonaws.services.s3.model.StaticEncryptionMaterialsProvider;
import com.amazonaws.services.s3.model.VersionListing;

public class S3ClientSideEncryptionAsymmetricMasterKey {
	private static final String keyDir = System.getProperty("java.io.tmpdir");
	private static final String bucketName = UUID.randomUUID() + "-"
			+ DateTimeFormat.forPattern("yyMMdd-hhmmss").print(new DateTime());
	private static final String objectKey = UUID.randomUUID().toString();

	public static void main(String[] args) throws Exception {

		// 1. Load keys from files
		byte[] bytes = FileUtils.readFileToByteArray(new File(keyDir
				+ "/private.key"));
		KeyFactory kf = KeyFactory.getInstance("RSA");
		PKCS8EncodedKeySpec ks = new PKCS8EncodedKeySpec(bytes);
		PrivateKey pk = kf.generatePrivate(ks);

		bytes = FileUtils.readFileToByteArray(new File(keyDir + "/public.key"));
		PublicKey publicKey = KeyFactory.getInstance("RSA").generatePublic(
				new X509EncodedKeySpec(bytes));

		KeyPair loadedKeyPair = new KeyPair(publicKey, pk);

		// 2. Construct an instance of AmazonS3EncryptionClient.
		EncryptionMaterials encryptionMaterials = new EncryptionMaterials(
				loadedKeyPair);
	    AWSCredentials credentials = new BasicAWSCredentials("USWUXHGYZQYFYFFIT3RE",
                "MOJRH0mkL1IPauahWITSVvyDrQbEEIwljvmxdq03");	    
		AmazonS3EncryptionClient encryptionClient = new AmazonS3EncryptionClient(
				credentials,
				new StaticEncryptionMaterialsProvider(encryptionMaterials));
		Region usEast1 = Region.getRegion(Regions.US_EAST_1);
		encryptionClient.setRegion(usEast1);
		encryptionClient.setEndpoint("http://localhost:9000");
		
		final S3ClientOptions clientOptions = S3ClientOptions.builder().setPathStyleAccess(true).build();
	    encryptionClient.setS3ClientOptions(clientOptions);
		
		// Create the bucket
		encryptionClient.createBucket(bucketName);
		// 3. Upload the object.
		byte[] plaintext = "Hello World, S3 Client-side Encryption Using Asymmetric Master Key!"
				.getBytes();
		System.out.println("plaintext's length: " + plaintext.length);
		encryptionClient.putObject(new PutObjectRequest(bucketName, objectKey,
				new ByteArrayInputStream(plaintext), new ObjectMetadata()));

		// 4. Download the object.
		S3Object downloadedObject = encryptionClient.getObject(bucketName,
				objectKey);
		byte[] decrypted = IOUtils.toByteArray(downloadedObject
				.getObjectContent());
		Assert.assertTrue(Arrays.equals(plaintext, decrypted));
		System.out.println("decrypted length: " + decrypted.length);
		//deleteBucketAndAllContents(encryptionClient);
	}

	private static void deleteBucketAndAllContents(AmazonS3 client) {
		System.out.println("Deleting S3 bucket: " + bucketName);
		ObjectListing objectListing = client.listObjects(bucketName);

		while (true) {
			for (Iterator<?> iterator = objectListing.getObjectSummaries()
					.iterator(); iterator.hasNext();) {
				S3ObjectSummary objectSummary = (S3ObjectSummary) iterator
						.next();
				client.deleteObject(bucketName, objectSummary.getKey());
			}

			if (objectListing.isTruncated()) {
				objectListing = client.listNextBatchOfObjects(objectListing);
			} else {
				break;
			}
		}
		client.deleteBucket(bucketName);
	}
}
```

**Comment by harshavardhana** [comment]
https://gist.github.com/harshavardhana/bcbd15638e5d443a6158191b2084ce6b - updated here with AES examples as well, both approaches 

- Symmetric (AES)
- Asymmetric (RSA)

seems to work fine with Minio - Minio server doesn't need any change after all to support aws-sdk encryptionClients it works transparently. 

But when we generate a presignedURL it would be generated for an encrypted object. Now when downloading this object through `curl` we would get an encrypted object because we haven't deciphered the object yet and `curl` has no awareness of such details. 

Now for presignedURL to work with the encryption we would need a proxy server of sort which has access to the `master.key` the requests needs to go through this proxy so that the streaming decoding can happen.  Will update with more findings on the proxy approach by coming week.

**Comment by harshavardhana** [comment]
Server already supports this, moving this issue into a documentation bug.

**Comment by GarimaKapoor** [comment]
@koolhead17 did you get a chance to test and document it?

**Comment by harshavardhana** [comment]
> @koolhead17 did you get a chance to test and document it?

ping @koolhead17

**Comment by GarimaKapoor** [comment]
@koolhead17 what is the timeline for this doc?

**Comment by koolhead17** [comment]
@GarimaKapoor I will be done with this by today.

**Comment by koolhead17** [comment]
Doc has been shared with @NitishT for final review & will be merged in cookbook.

**Comment by nitisht** [comment]
PR created : https://github.com/minio/cookbook/pull/142

**Comment by nitisht** [comment]
Fixed by minio/cookbook#142
